Doc 11873 3.2 GA

ElliotFrancisHunter opened the pull request:
This is a  PR for CBL 3.2 GA. Further changes including additional copy are coming soon.

Preview link: https://preview.docs-test.couchbase.com/cblga/home/index.html

Credentials: https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

P0 list: https://docs.google.com/document/d/1N1H3n4sCgq7Fuaj1ghVZX9ietHOuzYFU9Vg1frwOG7M/edit?usp=sharing

Changed Pages (Also available on the other main platforms, just linking swift as an example).
-------------------

- What's new: https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/cbl-whatsnew.html
- Vector Search (Swift): https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/vector-search.html
- Working with Vector Search(Swift): https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/working-with-vector-search.html
- Release Notes (Swift): https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/releasenotes.html
- Vector Search Release Notes (Swift): https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/vs-releasenotes.html
- Platform Compatibility: https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/compatibility.html
- Supported Environments: https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/supported-os.html

----------------------

Pages still needing changes/updates
-----------------------

-----------------------

ElliotFrancisHunter committed:
DOC:11873: Remove beta artefacts

Remove beta version tag from yaml. Remove beta copy from what's new and condense to one preamble about vector search

ElliotFrancisHunter committed:
DOC-11873: 3.2 GA updates

Change list:
-- Added updated supported formats for vector embeddings
-- Updated dimensions upper bounds
-- Update configuration table

pasin commented on modules/swift/pages/working-with-vector-search.adoc:
```diff
@@ -72,15 +72,22 @@ A document property for embedded vectors or
 |Optional
-|Euclidean Distance
-|You can also configure Cosine Distance as your Distance Metric
+|Euclidean Distance (euclideanSquared)
```
Squared Euclidean Distance

ElliotFrancisHunter replied:
Done!

ElliotFrancisHunter committed:
DOC-11873: V1 of new vector search doc

bmeike reviewed (commented):
I read only "New in 3.2" for Android and Java.
## Everywhere:
- There is a '{tag}" in all of the versions for all of the dependencies.  It should just go away: there will be no suffix

## Java:
If you want to use Vector Search, add the Couchbase Lite Vector Search dependency: com.couchbase.lite:couchbase-lite-java-vector-search-arm64-1.0.0{tag}: There is no such library.  We support several architectures: x86 and Apple silicon on MacOS, x86_64 on Linux, and x86_64 and ARM-64 on Windows.  The lib is named couchbase-lite-java-vector-search-1.0.0

## Android
- On this page: https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/android/gs-install.html
    - Kotlin - Enterprise: I suggest reversing steps 2 and 3
    - Kotlin - Enterprise: I suggest that #4 read "If you want to add Vector Search". 1) For x86_64, use <blah blah>-vector-search-x86_64-1.0.0. 2) for ARM-64 architectures use <blah blah>-vector-search-arm64-1.0.0
 - On this page: https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/android/working-with-vector-search.html
    - Create a Vector Index: I suggest "The *number* of centroids" in the second point
    - Vector Index Configuration: Expression => Further information.  I think that you are missing a disjunction between the two sentences
    - Vector Index Configuration: Distance Metric => Further information. There are 4 distance metrics: Euclidean Distance, Euclidean Squared, Cosine and Dot product
    - Generating Vectors: In point #1, I recommend emphasizing that you can use an *external* ML model... possible one on the far end of a network connection
    - Create a Vector Index with Embeddings: the snippet uses colors; the text says "words"
    - Use vector_distance(): did I not tag this correctly?

ElliotFrancisHunter committed:
DOC-11873: Vector Search GA updates for swift

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-11873-3.2-GA

ElliotFrancisHunter committed:
DOC-11873: V1 of updated vector search copy. All platforms

velicuvlad commented on modules/swift/examples/code_snippets/VectorSearch.swift:
```diff
@@ -66,8 +66,9 @@ class VectorSearchSnippets {
         config.numProbes = 8
         config.minTrainingSize = 2500
         config.maxTrainingSize = 5000
-        // end::vs-create-custom-config[]
```
The section name, where this tag is used, is [Create a Vector Index](https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/swift/working-with-vector-search.html#create-a-vector-index), contrary to the tag name. So it somewhat makes sense to have the call for createIndex within it if we go by section name, so I don't mind this change, but this should be reflected for all the other platforms as well for consistency. 

Now, for everything else than iOS - where for Obj-C needs only the tag to be moved, the `createIndex` call doesn't exist at all, so it will have to be added. That's more code changes needed. Thus, depending on the urgency, reverting this, Swift only, change might be preferred (moving the tag back in its original place).

CC: @idulo

ElliotFrancisHunter replied:
Reverted the tag for now. Will revisit the snippets wholesale after GA.

velicuvlad commented on modules/swift/examples/code_snippets/VectorSearch.swift:
```diff
@@ -66,8 +66,9 @@ class VectorSearchSnippets {
         config.maxTrainingSize = 5000
-        // end::vs-create-custom-config[]
+       
         try collection.createIndex(withName: "colors_index", config: config)
```
only exists for iOS snippets and I've addded it (outside of the tags) for validation purposes only.

ElliotFrancisHunter replied:
Noted. sorted tags

ElliotFrancisHunter committed:
DOC-11873: Remove 3.1.x relnotes

Merge artefacts

ElliotFrancisHunter committed:
DOC-11873: Relnote update

Release notes for CBL are now split into CBL specific and Vector Search specific release notes per platform. This is to improve visibility of both. Especially as the core CBL product is likely to be updated more regularly.

This has been reflected in the nav, the quickstart skeleton and anywhere you'd expect to find release notes.

As of yet there is no direct linking from CBL to Vector Search release notes but it's easy to find them.

Decision from PM to retain beta release notes as they were public releases.

ElliotFrancisHunter committed:
DOC-11873: Add platform specific relnotes

ElliotFrancisHunter committed:
DOC-11873: Remove errant beta mentions

Turns out tag: {empty} hides the tag attribute when not in use. Best to add that back in when there is no value for releasetag.

Also included updates to compatibility

osfameron commented on modules/android/pages/releasenotes.adoc:
```diff
@@ -10,3 +10,5 @@ include::partial$_set_page_context_for_android.adoc[]
 [#maint-latest]
 include::partial$release-notes/couchbase-mobile-android-release-note.3.2.0.adoc[]
+
+include::partial$release-notes/couchbase-mobile-android-release-note.3.2.0-beta.adoc[]
```
why did we decide to include the -beta relnotes again?

ElliotFrancisHunter replied:
They were public releases so it was decided to retain them for visibility. PM ask

osfameron commented on modules/android/partials/release-notes/couchbase-mobile-android-release-note.3.2.0-beta.adoc:
```diff
@@ -0,0 +1,202 @@
+[#maint-3-2-0-beta-1]
+== 3.2.0 Beta 1 -- March 2024
```
Does this still need to say "Beta 1"?
It looks like that essentially *was* the beta for Android? e.g. the beta versions just specify which mobile SDKs are involved?

osfameron commented on modules/android/partials/release-notes/couchbase-mobile-android-release-note.3.2.0-beta.adoc:
```diff
@@ -0,0 +1,202 @@
+
+=== Enhancements
+
+* https://issues.couchbase.com/browse/CBL-4378[CBL-4378 - Add consumer-rules.pro to Android maven and zip distributions]
```
This change is mentioned for the main 3.2.0 relnotes too, isn't it?

As both are included, that means that the bug will appear twice on the page?
This seems a bit odd to me.

ElliotFrancisHunter replied:
Good spot, will remove it from the GA notes if it's already present

osfameron commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -154,69 +153,139 @@ include::csharp:example$code_snippets/VectorSearch.cs[tags=vs-create-predictive-
+|image:ROOT:yes.png[]
+|The expression returning a vector (NOT Index Name).
+Must match the expression specified in the vector index.
+|target-vector
```
For all these tables, I would add a newline before each row, so you have some hope of figuring out what each cell is

osfameron commented on modules/objc/pages/working-with-vector-search.adoc:
```diff
@@ -21,69 +21,65 @@ NOTE: You must install Couchbase Lite to use the Vector Search extension.
 ----
 
-First, initialize the `config` variable with the `VectorIndexConfiguration()` method with the following parameters:
+First, initialize the `config` object with the `VectorIndexConfiguration()` method with the following parameters:
```
A question: how did you find the process of having to make the same changes multiple times for each SDK?

e.g. what's the trade-off with the simpler editing than common page-partials vs redoing the same work in multiple places for you?

RayOffiah replied:
Good question.

ElliotFrancisHunter replied:
For the most part the tradeoff is better visibility with less obfuscation vs risk of inconsistency. In the future I'd probably move the content in vector-search.adoc into a common partials file. However I find in practice that the working with page has too many small differences due to mobile SDK quirks that need calling out

osfameron reviewed (commented):
Looks like good work, I'm a bit confused about the -beta relnotes, but guess this is a PM request?

RayOffiah commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+=== About Lazy Vector Indexes
+
+You can use Lazy Vector Indexes(lazy index) to use a ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.
+Lazy indexing is an asynchronous process that provides developers full control during the following:
```
Not sure this reads correctly. I think 'during the following` doesn't match up with the list below.; it's not explaining during what.

ElliotFrancisHunter replied:
Edited from previous feedback

RayOffiah commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+
+NOTE: Updating in lazy index is an independent process from saving document operations.
+
+You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.
```
```suggestion
You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process,  and retrying the index update process at a later point when the ML model is available.
```

ElliotFrancisHunter replied:
Content removed and condensed down

RayOffiah commented on modules/android/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
 
+== Hybrid Vector Search
+
+You can use Vector Search with regular SQL++ query language. 
```
```suggestion
You can use Vector Search with the regular SQL++ query language. 
```

ElliotFrancisHunter replied:
Changed for a better explanation of the topic.

RayOffiah commented on modules/android/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
+
+You can use Vector Search with regular SQL++ query language. 
+Vector search will be performed on the documents that have been filtered based on the criteria specified in the WHERE clause.
+No LIMIT clause is required for Hybrid Vector Search.
```
```suggestion
The LIMIT clause is not required for Hybrid Vector Search.
```

RayOffiah commented on modules/android/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
+
+See the https://www.couchbase.com/blog/hybrid-search/[Hybrid Search blog post] for more information about Hybrid Search.
+
+IMPORTANT: You cannot `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```
```suggestion
IMPORTANT: You cannot  use `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```

ElliotFrancisHunter replied:
Done, moved to working with page

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+
+You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
```
Suggest you be more explicit about what proces you mean here. Do you mean Updating a lazy index?  Also consider combining this info with the previous note

ElliotFrancisHunter replied:
Actioned

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -181,7 +173,10 @@ include::java:example$snippets/common/main/java/com/couchbase/codesnippets/Vecto
 ----
 
-The snippet demonstrates the creation of the vector index, `word_index`, in the collection `_default.words`, demonstrating how you can create a vector index with a `word` embedding for use in queries related to written language.
+. First, create the standard configuration, setting up an expressions, number of dimensions and number of centroids for the vector embedding.
```
```suggestion
. First, create the standard configuration, setting up an expression, number of dimensions and number of centroids for the vector embedding.
```

ElliotFrancisHunter replied:
Good spot!

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -195,13 +190,17 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
 ----
 
-The snippet demonstrates the creation of the vector index, `word_index`, in the collection `_default.words`, demonstrating how you can create a vector index with a `word` embedding for use in queries related to written language.
+. First, create the standard configuration, setting up an expressions, number of dimensions and number of centroids for the vector embedding.
```
```suggestion
. First, create the standard configuration, setting up an expression, number of dimensions and number of centroids for the vector embedding.
```

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+
+=== About Lazy Vector Indexes
+
+You can use Lazy Vector Indexes(lazy index) to use a ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.
```
I don't think you need to i-cap lazy vector index....this is a standard industry term, not a branded feature.

also, spell out machine learning (ML)

ElliotFrancisHunter replied:
Done

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -233,32 +232,133 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
 
-Couchbase Lite currently supports two SQL++ functions, `vector_match()` and `vector_distance()`.
+Lazy indexing is an alternate approach to using the standard predictive model with regular vector indexes which handle the indexing process automatically.
+You can use lazy indexing to use a ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.
```
I could be wrong about this. 
```suggestion
You can use lazy indexing to use an ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.
```

ElliotFrancisHunter replied:
I get what you mean. I'll leave as is for now as I _think_ a machine learning model reads correctly compared to an ❓ 😖

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+=== About Lazy Vector Indexes
+
+You can use Lazy Vector Indexes(lazy index) to use a ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.
+Lazy indexing is an asynchronous process that provides developers full control during the following:
```
Lazy indexing is an asynchronous process that provides developers full control of:
* When to update the index
* The number of vectors to update to the index
* Whether to cancel or skip certain indexes when the ML model is unavailable or has failed.

ElliotFrancisHunter replied:
Much better, thanks! Changed.

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -233,32 +232,133 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
+--
+=====
+
+NOTE: Lazy Vector Indexing is opt in functionality, the `isLazy` property is set to `false` by default.
```
```suggestion
NOTE: Lazy Vector Indexing is opt-in functionality, the `isLazy` property is set to `false` by default.
```

ElliotFrancisHunter replied:
Good spot, hyphenation is always a sticking point for me

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -233,32 +232,133 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
+
+----
+
+You procedurally update the vectors in the index by looping through the vectors in batches equivalent to the `limit` parameter.
```
```suggestion
You procedurally update the vectors in the index by looping through the vectors in batches until you reach the value of the `limit` parameter.
```

RayOffiah commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -233,32 +232,133 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
+
+----
+
+You procedurally update the vectors in the index by looping through the vectors in batches equivalent to the `limit` parameter.
```
See previous comment

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+
+NOTE: Updating in lazy index is an independent process from saving document operations.
+
+You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.
```
I don't really understand this sentence....are you saying:

Use lazy indexing to handle situations when the machine learning model is unreachable or not functioning correctly.  In these cases, you can choose to schedule an on-demand index update later, when the ML model is available.

ElliotFrancisHunter replied:
Made text more concise. Should be clearer now.

RayOffiah commented on modules/swift/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
+
+NOTE: Updating in lazy index is an independent process from saving document operations.
+
+You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.
```
```suggestion
You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process,  and retrying the index update process at a later point when the ML model is available.
```

RayOffiah commented on modules/swift/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
 
+== Hybrid Vector Search
+
+You can use Vector Search with regular SQL++ query language. 
```
```suggestion
You can use Vector Search with the regular SQL++ query language. 
```

RayOffiah commented on modules/swift/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
+
+See the https://www.couchbase.com/blog/hybrid-search/[Hybrid Search blog post] for more information about Hybrid Search.
+
+IMPORTANT: You cannot `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```
```suggestion
IMPORTANT: You cannot use `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```

RayOffiah commented on modules/swift/pages/working-with-vector-search.adoc:
```diff
@@ -146,74 +153,144 @@ include::swift:example$code_snippets/VectorSearch.swift[tags=vs-create-predictiv
+You can enable lazy vector indexing by setting the `isLazy` property to `true` in your vector index configuration.
 
-|===
+NOTE: Lazy Vector Indexing is opt in functionality, the `isLazy` property is set to `false` by default.
```
```suggestion
NOTE: Lazy Vector Indexing is opt-in functionality, the `isLazy` property is set to `false` by default.
```

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
+
+See the https://www.couchbase.com/blog/hybrid-search/[Hybrid Search blog post] for more information about Hybrid Search.
+
+IMPORTANT: You cannot `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```
I don't understand what you mean in the "important" note.

ElliotFrancisHunter replied:
Actioned, moved to working with page near relevant snippets for better context

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -122,9 +181,19 @@ This metric focuses on the alignment of two vectors, the similarity of direction
 == Next Steps
 
-Now that you're familiar with the core concepts of vector search, including vector embeddings, indexing, encoding, centroids, dimensions, and distance metrics, it's time to put your knowledge into practice and start building AI applications.
+Now that you're familiar with the core concepts of vector search, it's time to put your knowledge into practice and start building AI applications.
```
I would strike the Next Steps section entirely. I don't think it adds value.

ElliotFrancisHunter replied:
Done

RayOffiah commented on modules/swift/pages/working-with-vector-search.adoc:
```diff
@@ -146,74 +153,144 @@ include::swift:example$code_snippets/VectorSearch.swift[tags=vs-create-predictiv
+
+----
+
+You procedurally update the vectors in the index by looping through the vectors in batches equivalent to the `limit` parameter.
```
```suggestion
You procedurally update the vectors in the index by looping through the vectors in batches until you reach the value of the `limit` parameter.
```

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -97,13 +150,19 @@ Choosing Centroids in vector search involves trade-offs that can impact clusteri
 
 The general guideline for the amount of Centroids is approximately the square root of the number of documents.
 
+=== About Probes
```
This is a little jarring--I don't know why you re mentioning probes....start with a statement to let user know what these are

ElliotFrancisHunter replied:
This is now addressed with clearer explanation surrounding probes.

RayOffiah reviewed (dismissed):
One or two minor points.

bfavini commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -30,101 +30,39 @@ Intel processors that support the Advanced Vector Extensions 2 (AVX2) instructio
 
-A beta version of Vector Search is now available on Couchbase Lite for the C platform.
+Vector Search is now available on Couchbase Lite for all platforms.
+Vector search is a sophisticated data retrieval technique that focuses on matching the contextual meanings of search queries and data entries, rather than simple text matching.
```
Vector Search

ElliotFrancisHunter replied:
Great spot!

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
 Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
 You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
 Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 
 Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.
```
Passive voice here makes it unclear as to who actually odes the configuration. Can you rephrase to make this clearer and simpler? Is this something I need to do? If not, how does it happen?

ElliotFrancisHunter replied:
It's overly wordy I admit as I copied verbatim. Did some research into the object and how it's stored. Likes like you can encode floats into a string format to store or transmit them in text-based formats, like JSON.

ElliotFrancisHunter replied:
Simplified

bfavini commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,65 @@ This representation encodes how a ML model understands the input or inputs provi
 You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
 Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 
 Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.
+If training is triggered during a query, the query may take longer to return results.
```
Be aware that vector index training can affect query performance.

I'd move the note about the required number of vectors up, directly after you introduce the concept of training.

ElliotFrancisHunter replied:
Done

ElliotFrancisHunter committed:
DOC-11873: P0 Updates

All P0 items identified by PM

osfameron committed:
release/3.2 compatibility updates

* Add 3.2 entry to table 1 "Sync Gateway and Couchbase Lite Compatibility Matrix"
* fix offset error in table 2 "Operating System SDK Support"

osfameron committed:
release/3.2 supported platform matrix updates

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-11873-3.2-GA

ElliotFrancisHunter committed:
DOC-11873: Added remaining snippets into .NET

Important to note is the .NET page is different. This is an experiment to deliver more useful copy. Post GA I may convert all of them to this format or change it so they're all the same.

osfameron committed:
tweaks after review

via Priya, Vlad on Slack

osfameron committed:
clarification

ElliotFrancisHunter committed:
DOC-11873: typo fixes

simon-dew commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -30,101 +30,39 @@ Intel processors that support the Advanced Vector Extensions 2 (AVX2) instructio
-
 ==== Extended Date/Time Functionality
 
 Six new DateTime functions have been added to Couchbase Lite N1Q1:
```
I know it's not part of this change, but should this be `{sqlpp}` or at least `N1QL`?

ElliotFrancisHunter replied:
Changed to N1QL

simon-dew commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
```
spacing

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
```

ElliotFrancisHunter replied:
Done

simon-dew commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
```
Minor spacing and punctuation edits

```suggestion
* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
```

ElliotFrancisHunter replied:
Done

simon-dew commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
Minor punctuation edits

```suggestion
* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

ElliotFrancisHunter replied:
Done

simon-dew commented on modules/c/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
Same minor spacing and punctuation updates as for android

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

simon-dew commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
Same minor punctuation and spacing suggestions as for android

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

simon-dew commented on modules/java/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
Same minor spacing and punctuation suggestions as for android

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

simon-dew commented on modules/objc/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
As for android

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

simon-dew commented on modules/swift/pages/vector-search.adoc:
```diff
@@ -58,12 +58,69 @@ This representation encodes how a ML model understands the input or inputs provi
+
+IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 
+
+Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
+
+* Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
+
+* The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```
Same as for android

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

simon-dew reviewed (commented):
Very quick once-over. Just a couple of minor suggestions

bmeike commented:
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/gs-install.html#introduction
* If you want to use Vector Search, add the Couchbase Lite Vector Search dependency: com.couchbase.lite:couchbase-lite-java-vector-search-arm64-1.0.0
There is no such package.  The Java vector search package is couchbase-lite-java-vector-search
## [couchbase-lite-java-vector-search](https://proget.sc.couchbase.com/feeds/internalmaven/com.couchbase.lite/couchbase-lite-java-vector-search/versions)
This is not correct.  The product binaries are not available from packages.couchbase.com.  If you need the support library, it is *ONLY* available from packages.couchbase.com
## LD_LIBRARY_PATH=<lib-directory>/support/libC++:<lib-directory>/support/libicu:$LD_LIBRARY_PATH
This is incorrect.  It should read: LD_LIBRARY_PATH="$LD_LIBRARY_PATH:<lib-directory>"
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/gs-install.html#using-gradle
This is inconsistent: required JDK 11 for Maven and 8 for Gradle.  Probably should be 11 for both
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/database.html#open-db
Application sandbox is an Android concept.  A Java db will be created in the current directory.
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/database.html#database-full-sync
"Database Full Sync is a safe method to preserve data loss but will incur "???  Why would I want to preserve data loss?
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/database.html#lbl-find-db-loc
"You should always use it to explicitly set the database location": this is not true.  There is no reason to set the database location unless you have a reason to do so.
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/working-with-vector-search.html#updating-the-lazy-index
I think I have a *much* better way of demonstrating this, now that I see it in print.  I'm going to update the example and submit a PR
## https://preview.docs-test.couchbase.com/cblga/couchbase-lite/current/java/working-with-vector-search.html#prediction-with-approx_vector_distance
I think it might be interesting to include an example that uses APPROX_VECTOR_DISTANCE in the SELECT clause.  That is, certainly, something one might do...

bmeike reviewed (dismissed):
There are, actually, a couple of things that really do need to be fixed.  They are probably just leftovers from ancient history...

bmeike commented:
Here is the PR for the improved Lazy VS example: https://github.com/couchbase/docs-couchbase-lite/pull/903

pasin reviewed (dismissed):
Quick comment on the installation page for C and iOS.

C

1. Vector Search for MacOS has a saparate debug symbol zip file from the binary zip file. The file name is `couchbase-lite-vector-search-1.0.0-macos-symbols.zip`

iOS

1. Both Objective-C and Swift Couchbase Lite 3.2.0 will support Carthage. So we need to add a Cathage tab on the installation page.The instruction can be copied from 3.1 and change the version number to 3.2.0.

2. Both Objective-C and Swift Vector Search Extension 1.0.0 will support Carthgae as well. The URL to that Carthage file is `http://packages.couchbase.com/releases/couchbase-lite-vector-search/carthage/CouchbaseLiteVectorSearch.json`

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-11873-3.2-GA

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-11873-3.2-GA

ElliotFrancisHunter committed:
DOC-11873: Apply dev feedback.

ElliotFrancisHunter committed:
DOC-11873: V3 doc fixes

Change Summary:
-- Clarifying language around supported formats, lazy indexes and training
-- SPG fixes in working with pages

ElliotFrancisHunter committed:
DOC-11873: .NET install correction

Correctly reference android activation in android section

ElliotFrancisHunter committed:
DOC-11873: Add missing iOS ticket

ElliotFrancisHunter committed:
DOC-11873: Correcting JDK mismatch. All v11 now

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-11873-3.2-GA

borrrden commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -133,16 +134,14 @@ include::csharp:example$code_snippets/VectorSearch.cs[tags=vs-create-index]
 
 This method generates vectors to be indexed for each document at the index time by using the `prediction()` function.
-The key difference to note is that the `config` object uses the output of the `prediction()` function as the `expression` parameter to generate the color embedding.
+The key difference to note is that the `config` object uses the output of the `prediction()` function as the `expression` parameter to generate the vector index.
```
This section has a snippet following it that doesn't really tell the whole story.  The user needs some guidance on the way that prediction function itself interacts with the color model (it's similar, if not identical, to predictive query).  That is, that the `colorInput` comes from a document property into the prediction function, and `vector` is a key on the result dictionary that is written by the prediction function and that is why the syntax `{\"colorInput\": color}).vector` is used.  The generic naming of `vector` could mislead people into thinking it is just a magic name.

ElliotFrancisHunter replied:
Addressed

ElliotFrancisHunter committed:
DOC-11873: Add .NET syntax explanation for prediction

ElliotFrancisHunter committed:
DOC-11873: Added clearer explanation for .NET specific syntax

osfameron commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -30,104 +30,42 @@ Intel processors that support the Advanced Vector Extensions 2 (AVX2) instructio
 ==== Extended Date/Time Functionality
 
-Six new DateTime functions have been added to Couchbase Lite N1Q1:
+Six new DateTime functions have been added to Couchbase Lite N1QL:
```
SQL++... ?

osfameron commented on modules/android/pages/vector-search.adoc:
```diff
@@ -58,12 +58,74 @@ This representation encodes how a ML model understands the input or inputs provi
+
+|Update when documents are changed
+|image:ROOT:yes.png[]
+|image:ROOT:no.png[]
```
Having spent some "quality time" recently with tables... please put a new-line in between rows, for sanity's sake 😂

osfameron commented on modules/c/pages/vector-search.adoc:
```diff
@@ -1,12 +1,12 @@
 = Vector Search
-:page-status: Beta
+:page-status:
```
I suspect this should be commented out rather than left blank, as (weirdly), blank is a "truthy" value.
But I saw you had some back and forth on this earlier, so not important while the release needs to go ahead, can look at it together in a tidy-up phase later 👍

simon-dew commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -181,7 +173,10 @@ include::java:example$snippets/common/main/java/com/couchbase/codesnippets/Vecto
-The snippet demonstrates the creation of the vector index, `word_index`, in the collection `_default.words`, demonstrating how you can create a vector index with a `word` embedding for use in queries related to written language.
+. First, create the standard configuration, setting up an expression, number of dimensions and number of centroids for the vector embedding.
+
+. Next, create a vector index, `colors_index`, on a collection and pass it our configuration.
```
Style rule: avoid first person

```suggestion
. Next, create a vector index `colors_index` on a collection, and pass the configuration to it.
```

ElliotFrancisHunter replied:
Will address these post GA

simon-dew commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -195,13 +190,17 @@ include::example$kotlin_snippets/app/src/main/kotlin/com/couchbase/codesnippets/
-The snippet demonstrates the creation of the vector index, `word_index`, in the collection `_default.words`, demonstrating how you can create a vector index with a `word` embedding for use in queries related to written language.
+. First, create the standard configuration, setting up an expression, number of dimensions and number of centroids for the vector embedding.
+
+. Next, create a vector index, `colors_index`, on a collection and pass it our configuration.
```
Avoid "our" again

ElliotFrancisHunter replied:
Resolved these

simon-dew commented on modules/android/pages/working-with-vector-search.adoc:
```diff
@@ -142,26 +124,36 @@ Reduces the number of bits per dimension
 
-* The minimum training size is set to 25x the number of Centroids
+* The minimum training size is set to 25x the number of Centroids or 2 ^PQ's^ ^bits^ when PQ is used
 
-* The maximum training size is set to 256x the number of Centroids
+* The maximum training size is set to 256x the number of Centroids or 2 ^PQ's^ ^bits^ when PQ is used
```
Looking at the preview, I don't think it's immediately obvious with the raised text, so maybe spell it out "2 to the power of PQ's bits"?

simon-dew reviewed (commented):
Had a quick scan through and all my previous comments have been addressed. Just found a couple of minor stylistic points in the new edits.

ElliotFrancisHunter committed:
DOC-11873: Fix binary download links

ElliotFrancisHunter committed:
DOC-11873: Quick wins for platform support
